### DISCUSSION TITLE
'concourse web' fails: panic: runtime error: index out of range [-1]

### DISCUSSION THREAD

**Discussion by yurivict** [concourse/concourse#6081] [Help & Support]
I followed the tutorial https://concourse-ci.org/concourse-generate-key.html and got this failure.
Perhaps some key is mixed up, but concourse should say what is wrong instead of panicking.

```
{"timestamp":"2020-09-21T05:46:58.836389518Z","level":"info","source":"atc","message":"atc.cmd.start","data":{"session":"1"}}
{"timestamp":"2020-09-21T05:46:58.836667757Z","level":"info","source":"atc","message":"atc.cmd.finish","data":{"duration":92667,"session":"1"}}
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/concourse/concourse/atc/db/migration.(*migrator).Up(0xc000448050, 0x99d480, 0xc000618200)
	/usr/ports/misc/concourse/work/concourse-6.5.1/atc/db/migration/migration.go:370 +0xa5
github.com/concourse/concourse/atc/db/migration.(*OpenHelper).Open(0xc00079ed28, 0x0, 0x0, 0x0)
	/usr/ports/misc/concourse/work/concourse-6.5.1/atc/db/migration/migration.go:63 +0xb9
github.com/concourse/concourse/atc/db.Open(0x9ff920, 0xc0000a8f00, 0x71ad55, 0x1d, 0xc0000d4780, 0x71, 0x0, 0x0, 0x6e5107, 0x3, ...)
	/usr/ports/misc/concourse/work/concourse-6.5.1/atc/db/open.go:77 +0x152
github.com/concourse/concourse/atc/atccmd.(*RunCommand).constructDBConn(0xc0001ce500, 0x71ad55, 0x1d, 0x9ff920, 0xc0000a8c60, 0xa, 0x6e5107, 0x3, 0x99d480, 0xc000618200, ...)
	/usr/ports/misc/concourse/work/concourse-6.5.1/atc/atccmd/command.go:1492 +0x1cd
github.com/concourse/concourse/atc/atccmd.(*RunCommand).Runner(0xc0001ce500, 0xc0003b7910, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0)
	/usr/ports/misc/concourse/work/concourse-6.5.1/atc/atccmd/command.go:492 +0x747
main.(*WebCommand).Runner(0xc0004ded88, 0xc0003b7910, 0x0, 0x1, 0x7, 0x6, 0xc0003b7780, 0xc000271cc0)
	/usr/ports/misc/concourse/work/concourse-6.5.1/cmd/concourse/web.go:62 +0xa5
main.(*WebCommand).Execute(0xc0004ded88, 0xc0003b7910, 0x0, 0x1, 0x447820, 0x6af9e0)
	/usr/ports/misc/concourse/work/concourse-6.5.1/cmd/concourse/web.go:44 +0x65
github.com/vito/twentythousandtonnesofcrudeoil.installEnv.func2(0x82ab53cd8, 0xc0004ded88, 0xc0003b7910, 0x0, 0x1, 0x1, 0x203000)
	/usr/ports/misc/concourse/work/concourse-6.5.1/vendor/github.com/vito/twentythousandtonnesofcrudeoil/environment.go:40 +0x8a
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc00019fb90, 0xc00000e090, 0x1, 0x1, 0xc000319680, 0x0, 0x0, 0x0, 0x0)
	/usr/ports/misc/concourse/work/concourse-6.5.1/vendor/github.com/jessevdk/go-flags/parser.go:340 +0x852
github.com/jessevdk/go-flags.(*Parser).Parse(...)
	/usr/ports/misc/concourse/work/concourse-6.5.1/vendor/github.com/jessevdk/go-flags/parser.go:190
main.main()
	/usr/ports/misc/concourse/work/concourse-6.5.1/cmd/concourse/main.go:30 +0x213
```

concourse-6.5.1

**Comment by jamieklassen** [comment]
Converted to a discussion per https://blog.concourse-ci.org/nobody-wants-a-stale-bot/#an-update-on-triage. @chenbh @aoldershaw at first I saw this and got worried that, since keys were mentioned, that #5980 might have caused it. But that PR was not released in 6.5.

@yurivict based on your bio, and the `/usr/ports` prefixes in those paths, I did a bit of digging. Looks like you're the author of https://svnweb.freebsd.org/ports/head/misc/concourse/ -- cool! From my experiments with FreeBSD as a teenager, I know the ports system involves building packages from source.

Most likely this problem arises because you're building concourse using `go build`, whereas [our ci uses `packr build`](https://github.com/concourse/ci/blob/5220339a6d90a8ebfba57f578eb694c93288eb3b/tasks/scripts/concourse-build#L37): https://github.com/gobuffalo/packr.

This theory is borne out by the evidence that your code is panicking on

https://github.com/concourse/concourse/blob/5edc00f848aa6c3659b25c1900e1be31a6c82b4e/atc/db/migration/migration.go#L370

which suggests that `migrations` is an empty slice. Those migrations are intended to be bundled with the `packr` CLI, so when you do a straight `go build` that list probably is empty.

Once I stared at it for a while, your issue reminded me of #5613 when @flokli was working on creating a concourse package for Nix, which also builds from source.

**Comment by yurivict** [comment]
> Most likely this problem arises because you're building concourse using go build ...

I use the FreeBSD ports' framework module "go" that builds from source using ```go build```.

**Comment by flokli** [comment]
This might all get a bit cleaner once https://github.com/golang/go/issues/41191 has worked its way into go - see https://go.googlesource.com/proposal/+/master/design/draft-embed.md for a rendered draft.
